internal/deployer/deployer.go

```diff
@@ -1,96 +1,140 @@
 )
 
-type deployer struct {
+// Deployer struct manages project deployment
+type Deployer struct {
 	tfPluginClient *gridDeployer.TFPluginClient
-}
 
-func NewDeployer() deployer {
-	mnemonics := os.Getenv("MNEMONICS")
-	log.Printf("mnemonics: %s", mnemonics)
+	logger zerolog.Logger
+}
 
-	network := os.Getenv("NETWORK")
-	log.Printf("network: %s", network)
+// NewDeployer return new project deployer
+func NewDeployer(mnemonics, network string, showLogs bool) (Deployer, error) {
+	rand.Seed(time.Now().UnixNano())
 
 	tfPluginClient, err := gridDeployer.NewTFPluginClient(mnemonics, "sr25519", network, "", "", true, false)
 	if err != nil {
-		return deployer{}
+		return Deployer{}, err
 	}
-	return deployer{tfPluginClient: &tfPluginClient}
-}
-
-func (b *deployer) Deploy(ctx context.Context, repoURL string) (string, error) {
-	network := workloads.ZNet{
-		Name:        "gridifyNetwork",
-		Description: "network for testing",
-		Nodes:       []uint32{3},
-		IPRange: gridtypes.NewIPNet(net.IPNet{
-			IP:   net.IPv4(10, 20, 0, 0),
-			Mask: net.CIDRMask(16, 32),
-		}),
-		AddWGAccess: false,
+	logger := zerolog.New(zerolog.ConsoleWriter{Out: os.Stderr}).Level(zerolog.InfoLevel).
+		With().
+		Timestamp().
+		Logger()
+	if !showLogs {
+		logger = zerolog.New(zerolog.ConsoleWriter{Out: io.Discard})
 	}
 
-	vm := workloads.VM{
-		Name:       "gridifyVM",
-		Flist:      "https://hub.grid.tf/aelawady.3bot/abdulrahmanelawady-gridify-test-latest.flist",
-		CPU:        2,
-		PublicIP:   true,
-		Planetary:  true,
-		Memory:     1024,
-		RootfsSize: 20 * 1024,
-		Entrypoint: "/sbin/zinit init",
-		EnvVars: map[string]string{
-			"REPO_URL": repoURL,
+	return Deployer{
+			tfPluginClient: &tfPluginClient,
+			logger:         logger,
 		},
-		NetworkName: network.Name,
-	}
+		nil
+}
+
+// Deploy deploys a project and map each port to a domain
+func (d *Deployer) Deploy(ctx context.Context, repoURL, projectName string, ports []string) (map[string]string, error) {
 
-	err := b.tfPluginClient.NetworkDeployer.Deploy(ctx, &network)
+	randomString := randString(10)
+
+	d.logger.Info().Msg("getting nodes with free resources")
+	filter := gridDeployer.NodeFilter{
+		CRU:       2,
+		MRU:       2,
+		HRU:       5,
+		PublicIPs: true,
+		Gateway:   true,
+		Status:    "up",
+	}
+	nodes, err := gridDeployer.FilterNodes(filter, gridDeployer.RMBProxyURLs[d.tfPluginClient.Network])
 	if err != nil {
-		return "", err
+		return map[string]string{}, errors.Wrapf(
+			err,
+			"failed to get a node with enough resources on network %s",
+			d.tfPluginClient.Network,
+		)
+	}
+	if len(nodes) == 0 {
+		return map[string]string{}, fmt.Errorf(
+			"no available node with enough resources on network %s",
+			d.tfPluginClient.Network,
+		)
 	}
 
-	dl := workloads.NewDeployment("gridifyVM", 3, "", nil, network.Name, nil, nil, []workloads.VM{vm}, nil)
-	err = b.tfPluginClient.DeploymentDeployer.Deploy(ctx, &dl)
+	networkName := fmt.Sprintf("gnet%s", randomString)
```

Comment:
I thibk we can make a constrcutNetwork and add it to it
then we deploy outside function